### PR TITLE
DRC: math: Replace exponential function for performance

### PR DESCRIPTION
**Pull request by ShriramShastry**
For DRC performance, replace `exp_small_fixed()` with `sofm_exp_int32()`.
Included supporting change to include `sofm_exp_int32()` within `exp_fixed()`
and repositioned `exp_small_fixed()` for future use.

### CODE REVIEW

**Comment by ShriramShastry** [comment]
For compilation, this PR requires further fixes.

**Comment by singalsu** [comment]
In my test this patch dropped drc.2.1 average MCPS from 140 to 96 in topology sof-hda-efx-generic-4ch.tplg. Excellent!

The used topology initializes DRC to passthrough, so I applied another with "sof-ctl -n 36 -s threshold_-35_knee_27_ratio_8.txt". The blob that I used  is not in SOF git yet but I'll attach it here:

[threshold_-35_knee_27_ratio_8.txt](https://github.com/thesofproject/sof/files/13251190/threshold_-35_knee_27_ratio_8.txt)

**Review by singalsu — CHANGES_REQUESTED** [review]

File: src/math/decibels.c [hunk 1]
```diff
@@ -105,10 +72,10 @@ int32_t exp_fixed(int32_t x)
 		n++;
 	}
 
-	/* exp_small_fixed() input is Q3.29, while x1 is Q5.27
-	 * exp_small_fixed() output is Q9.23, while y0 is Q12.20
+	/* sofm_exp_int32() input is Q4.28, while x1 is Q5.27
+	 * sofm_exp_int32() output is Q9.23, while y0 is Q12.20
 	 */
-	y0 = Q_SHIFT_RND(exp_small_fixed(Q_SHIFT_LEFT(xs, 27, 29)), 23, 20);
+	y0 = Q_SHIFT_RND(sofm_exp_int32(Q_SHIFT_LEFT(xs, 27, 28)), 23, 20);
```
Comment (btian1) [inline src/math/decibels.c:78]: line 78 - 81 still can be optimized further with instrinsic code.

Comment (andrula-song) [reply src/math/decibels.c:78]: this file didn't include xtensa header file, do you mean add hifi implementation of functions in this file?

Comment (btian1) [reply src/math/decibels.c:78]: yes, each for loop deserve an intrinsic implementation.

**Comment by lgirdwood** [comment]
@singalsu @ShriramShastry @andrula-song I think we need to have some options where we can decide at build time via Kconfig between speed and accuracy for certain maths ops. i.e. for 16bit output we may not need full 24/32 computations and so on. There may also be usages where speed is more impartnat than bit accuracy, but thats up to you guys to identify and target.

File: src/audio/Kconfig [hunk 2]
```diff
@@ -280,6 +280,7 @@ config COMP_DRC
 	select NUMBERS_NORM
 	select MATH_DECIBELS
```
[the remark this answers is not in the document]
Comment (singalsu) [reply src/audio/Kconfig:281]: I meant don't remove or edit MATH_DECIBELS function but add the needed functions to MATH_EXP. Then you can remove from Kconfig for DRC the dependency to MATH_DECIBELS.

File: src/math/decibels.c (hunk 1 above)
Comment (singalsu) [reply src/math/decibels.c:78]: Please check if it improves speed. In some cases the C multiply has been as fast.

**Comment by singalsu** [comment]
> @singalsu @ShriramShastry @andrula-song I think we need to have some options where we can decide at build time via Kconfig between speed and accuracy for certain maths ops. i.e. for 16bit output we may not need full 24/32 computations and so on. There may also be usages where speed is more important than bit accuracy, but thats up to you guys to identify and target.

Yep, I wanted to avoid Sriram to change the old exponent function code since it is used by many other features. The new version for DRC is significantly faster but trades off slightly accuracy. We need to check case by case with exponent function to utilize.

File: src/math/decibels.c (hunk 1 above)
Comment (ShriramShastry) [reply src/math/decibels.c:78]: I'll give { `Q_CONVERT_FLOAT` ~, { `Q_SHIFT_RND` ~, { `Q_SHIFT_LEFT` ~, and `Q_MULTSR_32X32` } a shot as well. If I don't succeed, I'll try again with the next PR.

**Comment by singalsu** [comment]
@ShriramShastry Any updates to this? The MCPS saving is large so we should get this cleaned up and merged.

File: src/math/decibels.c (hunk 1 above)
Comment (ShriramShastry) [reply src/math/decibels.c]: image

Comment (ShriramShastry) [reply src/math/decibels.c]: I have completed the implementation of Q_SHIFT_RND, Q_SHIFT_LEFT, and Q_MULTSR_32X32, and I am currently working on Q_CONVERT_FLOAT.

**Review by lgirdwood — APPROVED** [review]
LGTM. @andrula-song pls review.

**Comment by lgirdwood** [comment]
@ShriramShastry some conflicts, pls rebase as CI wont complete.

File: src/math/decibels.c (hunk 1 above)
Comment (ShriramShastry) [reply src/math/decibels.c]: I'll correct the last few CI errors.

File: src/math/exp_fcn.c [hunk 3]
```diff
@@ -217,4 +221,52 @@ int32_t sofm_exp_int32(int32_t x)
...
+ */
+int32_t sofm_exp_fixed(int32_t x)
+{
+	int32_t xs;
+	int32_t y;
+	int32_t y0;
+	int i;
+	int n = 0;
+
+	if (x < Q_CONVERT_FLOAT(-11.5, 27))
```
Comment (andrula-song) [inline src/math/exp_fcn.c:249]: can we use a macro replace those fixed float convert? from my test, the division really costs many cycles.

File: src/math/exp_fcn_hifi.c [hunk 4]
```diff
@@ -278,4 +284,130 @@ int32_t sofm_exp_int32(int32_t x)
...
+ */
+int32_t sofm_exp_fixed(int32_t x)
+{
+	int32_t xs;
+	int32_t y;
+	int32_t y0;
+	int i;
+	int n = 0;
+
+	if (x < XT_Q_CONVERT_FLOAT(-11.5, 27))
```
Comment (andrula-song) [inline src/math/exp_fcn_hifi.c:374]: here too

Comment (ShriramShastry) [reply src/math/exp_fcn_hifi.c:374]: The updated macro is now in use here.

File: src/math/exp_fcn.c (hunk 3 above)
Comment (ShriramShastry) [reply src/math/exp_fcn.c:249]: Yup, We can swap it out, saving about 93% of the cycles compared to Generic C. We can handle this in the future PR as the present one just addresses modifications for exponential. I hope you will consent to this.

**Review by andrula-song — APPROVED** [review]

File: src/math/exp_fcn.c (hunk 3 above)
Comment (andrula-song) [reply src/math/exp_fcn.c:249]: actually, I'd like use a simple macro like #define THERSHOULD Q_CONVERT_FLOAT(-11.5, 27) to replace the convert. because the result is a fixed statistic.

File: src/math/exp_fcn_hifi.c (hunk 4 above)
Comment (andrula-song) [reply src/math/exp_fcn_hifi.c:374]: curious about the cycles that XT_Q_CONVERT_FLOAT takes.

Comment (ShriramShastry) [reply src/math/exp_fcn_hifi.c:374]: https://github.com/thesofproject/sof/pull/8435#discussion_r1404095250

File: src/math/exp_fcn.c [hunk 5]
```diff
@@ -20,7 +23,7 @@
 #define SOFM_BIT_MASK_Q62P2 0x4000000000000000LL
 #define SOFM_CONVERG_ERROR 28823037607936LL	// error smaller than 1e-4,1/2 ^ -44.7122876200884
 #define SOFM_BIT_MASK_LOW_Q27P5 0x8000000
-#define SOFM_QUOTIENT_SCALE BIT(30)
+#define SOFM_QUOTIENT_SCALE 0x400000000
```
[the remark this answers is not in the document]
Comment (ShriramShastry) [reply src/math/exp_fcn.c:26]: Oops, your right !!

File: src/math/exp_fcn_hifi.c [hunk 6]
```diff
@@ -26,10 +27,13 @@
...
+#include <xtensa/tie/xt_FP.h>
+
 #define SOFM_CONVERG_ERROR 28823037624320LL /* error smaller than 1e-4,1/2 ^ -44.7122876209085 */
 #define SOFM_BIT_MASK_LOW_Q27P5 0x0000000008000000
 #define SOFM_BIT_MASK_Q62P2 0x4000000000000000LL
-#define SOFM_QUOTIENT_SCALE BIT(30)
+#define SOFM_QUOTIENT_SCALE 0x40000000
 #define SOFM_TERMS_Q23P9 0x800000
 #define SOFM_LSHIFT_BITS 0x2000
 /*
```
Comment (lgirdwood) [inline src/math/exp_fcn_hifi.c:39]: Btw, these macros look similar (the same ?) as macros in the previous source file. If so, we should declare them in a exp header.

File: src/math/exp_fcn_hifi.c [hunk 7]
```diff
@@ -278,4 +284,130 @@ int32_t sofm_exp_int32(int32_t x)
 
 	return AE_MOVAD32_L(AE_MOVINT32X2_FROMINT64(ts));
 }
+
+/* Xtensa macros */
+/* Convert a float number to fractional Qnx.ny format. Note that there is no
+ * check for nx+ny number of bits to fit the word length of int. The parameter
+ * qy must be 31 or less.
+ */
+static inline int sofm_q_CONVERT_FLOAT(float a, int b)
```
Comment (lgirdwood) [inline src/math/exp_fcn_hifi.c:293]: sorry, my mistake - I did not see these were in the hifi file and not public APIs. These can be renamed to ```exp_hifi_q_convert_float()``` all lower case and so on.

**Comment by keqiaozhang** [comment]
SOFCI TEST

**Comment by keqiaozhang** [comment]
SOFCI TEST

File: src/math/exp_fcn.c [hunk 8]
```diff
@@ -217,4 +216,52 @@ int32_t sofm_exp_int32(int32_t x)
 	}
 	return ts;
 }
+
+#define ONE_Q20         Q_CONVERT_FLOAT(1.0, 20)	  /* Use Q12.20 */
+#define TWO_Q27         Q_CONVERT_FLOAT(2.0, 27)	  /* Use Q5.27 */
+#define MINUS_TWO_Q27   Q_CONVERT_FLOAT(-2.0, 27)	  /* Use Q5.27 */
```
Comment (btian1) [inline src/math/exp_fcn.c:222]: those definition are too simple, suggest add EXP_ to avoid potential conflicts.
or you may directly use existing definition:
https://github.com/thesofproject/sof/blob/main/src/math/decibels.c

put it to a header file.

File: src/math/exp_fcn.c [hunk 9]
```diff
@@ -217,4 +216,52 @@ int32_t sofm_exp_int32(int32_t x)
 	}
 	return ts;
 }
+
+#define ONE_Q20         Q_CONVERT_FLOAT(1.0, 20)	  /* Use Q12.20 */
```
Comment (btian1) [inline src/math/exp_fcn.c:220]: duplicated name, suggest move to common header.

File: src/math/decibels.c (hunk 1 above)
Comment (ShriramShastry) [reply src/math/decibels.c]: I've finished addressing every CI error.

Comment (btian1) [reply src/math/decibels.c]: @ShriramShastry , still code style error.

File: src/math/decibels.c [hunk 10]
```diff
@@ -105,10 +106,10 @@ int32_t exp_fixed(int32_t x)
 		n++;
 	}
 
-	/* exp_small_fixed() input is Q3.29, while x1 is Q5.27
-	 * exp_small_fixed() output is Q9.23, while y0 is Q12.20
+	/* sofm_exp_int32() input is Q4.28, while x1 is Q5.27
```
Comment (btian1) [inline src/math/decibels.c]: for comments, you mentioned input is Q4.28, however, what is x1? means sofm_exp_int32() output?
if means output, please change x1 to output.

File: src/math/exp_fcn.c [hunk 11]
```diff
@@ -212,4 +213,54 @@ int32_t sofm_exp_int32(int32_t x)
...
+ * the input argument for private small value exp() function that is
+ * accurate with input range -2.0 .. +2.0. The number of possible
+ * divisions by 2 is computed into variable n. The returned value is
+ * exp()^(2^n).
+ *
+ * Input  is Q5.27, -16.0 .. +16.0, but note the input range limitation
+ * Output is Q12.20, 0.0 .. +2048.0
+ */
+
+int32_t exp_fixed(int32_t x)
```
Comment (btian1) [inline src/math/exp_fcn.c]: your patch sequence are quite strange, normally, we should have c version first, then hifi version, seems you already have exp_fixed hifi in previous patch, now this patch comes with c version, this is strange sequence.

File: src/math/exp_fcn_hifi.c [hunk 12]
```diff
@@ -302,10 +367,10 @@ int32_t exp_fixed(int32_t x)
 	int i;
 	int n = 0;
 
-	if (x < Q_CONVERT_FLOAT(-11.5, 27))
+	if (x < exp_hifi_q_convert_float(-11.5, 27))
```
Comment (btian1) [inline src/math/exp_fcn_hifi.c:367]: this is not build with pre-compiler stage? could you show the asm code to compare?

File: src/math/exp_fcn_hifi.c [hunk 13]
```diff
@@ -277,10 +280,72 @@ int32_t sofm_exp_int32(int32_t x)
...
+	ae_int32 xt_o;
+	int shift;
+
+	shift = XT_SUB(c, b);
+	xt_o = AE_SLAA32(a, shift);
+
+	return xt_o;
+}
+
+#define ONE_Q20         exp_hifi_q_convert_float(1.0, 20)	  /* Use Q12.20 */
```
Comment (btian1) [inline src/math/exp_fcn_hifi.c]: these are always repeating definition, please move to one common header.

File: src/math/exp_fcn_hifi.c [hunk 14]
```diff
@@ -318,10 +383,13 @@ int32_t exp_fixed(int32_t x)
...
 	 */
-	y0 = Q_SHIFT_RND(sofm_exp_int32(Q_SHIFT_LEFT(xs, 27, 28)), 23, 20);
+	y0 = exp_hifi_q_shift_rnd(
+							  sofm_exp_int32(
+							  exp_hifi_q_shift_left(xs, 27, 28)),
+							  23, 20);
 	y = ONE_Q20;
 	for (i = 0; i < (1 << n); i++)
-		y = (int32_t)Q_MULTSR_32X32((int64_t)y, y0, 20, 20, 20);
+		y = (int32_t)exp_hifi_q_multsr_32x32((int64_t)y, y0, 20, 20, 20);
```
Comment (btian1) [inline src/math/exp_fcn_hifi.c:389]: this patch is not related with title(exp), you can't use one PR to cover all the changes, please move out this patch from this PR.

File: src/math/window.c [hunk 15]
```diff
@@ -114,7 +115,7 @@ void win_povey_16b(int16_t win[], int length)
 		/* Calculate x^0.85 as exp(0.85 * log(x)) */
 		x2 = (int32_t)(ln_int32((uint32_t)x1) >> 1) - WIN_LOG_2POW31_Q26;
 		x3 = sat_int32(Q_MULTSR_32X32((int64_t)x2, WIN_085_Q31, 26, 31, 27)); /* Q5.27 */
-		x4 = exp_fixed(x3); /* Q5.27 -> Q12.20 */
+		x4 = sofm_exp_fixed(x3); /* Q5.27 -> Q12.20 */
```
Comment (btian1) [inline src/math/window.c:118]: somehow, you can rename from the beginning or start from first patch, then this patch will be removed.

File: zephyr/CMakeLists.txt [hunk 16]
```diff
@@ -370,6 +370,8 @@ zephyr_library_sources(
 	${SOF_MATH_PATH}/decibels.c
 	${SOF_MATH_PATH}/numbers.c
 	${SOF_MATH_PATH}/trig.c
+	${SOF_MATH_PATH}/exp_fcn.c
+	${SOF_MATH_PATH}/exp_fcn_hifi.c
```
Comment (marc-hb) [inline zephyr/CMakeLists.txt:374]: heads-up this will conflict with #8548

File: src/audio/module_adapter/module/generic.c [hunk 17]
```diff
@@ -198,7 +198,7 @@ int module_prepare(struct processing_module *mod,
 		   struct sof_source **sources, int num_of_sources,
 		   struct sof_sink **sinks, int num_of_sinks)
 {
-	int ret;
+	int ret = 0;
```
Comment (singalsu) [inline src/audio/module_adapter/module/generic.c:201]: I made without knowing about this separate PR #8605. You can leave this fix out from DRC PR. I will add credits of patch to you since you did this first.

File: src/math/decibels.c [hunk 18]
```diff
@@ -9,42 +9,8 @@
 #include <sof/audio/format.h>
 #include <stdint.h>
 
-#define ONE_Q23         Q_CONVERT_FLOAT(1.0, 23)	  /* Use Q9.23 */
 #define LOG10_DIV20_Q27 Q_CONVERT_FLOAT(0.1151292546, 27) /* Use Q5.27 */
 
-/* Exponent function for small values of x. This function calculates
```
Comment (singalsu) [inline src/math/decibels.c:17]: Don't remove this. Both exponent functions should be possible to enable via kconfig. To be able to select both until we can safely remove the slower they should have different function name. DRC should use this new fast function.

Can your exponent function have different name, like have _fast_ in function name or some algorithm specific keyword to differentiate?  Or rename exp_fixed() and all usages of it (except DRC)  to e.g. exp_taylor_fixed(), since it's using Taylor -series.

**Review by singalsu — CHANGES_REQUESTED** [review]

File: src/math/exp_fcn_hifi.c [hunk 19]
```diff
@@ -274,4 +275,54 @@ int32_t sofm_exp_int32(int32_t x)
...
+ * the input argument for private small value exp() function that is
+ * accurate with input range -2.0 .. +2.0. The number of possible
+ * divisions by 2 is computed into variable n. The returned value is
+ * exp()^(2^n).
+ *
+ * Input  is Q5.27, -16.0 .. +16.0, but note the input range limitation
+ * Output is Q12.20, 0.0 .. +2048.0
+ */
+
+int32_t exp_fixed(int32_t x)
```
Comment (singalsu) [inline src/math/exp_fcn_hifi.c]: You don't get conflict with the other exponent if you use here other name, e.g. sofm_exp_<your method name>_fixed(). And DRC should use the new function name.

File: src/math/exp_fcn_hifi.c (hunk 12 above)
Comment (ShriramShastry) [reply src/math/exp_fcn_hifi.c:367]: Sorry, I don't get the necessity in this situation; could you please explain?

File: src/math/exp_fcn_hifi.c [hunk 20]
```diff
@@ -277,10 +280,72 @@ int32_t sofm_exp_int32(int32_t x)
...
+	x = XT_FLOAT_S(val, 0);
+	y = XT_MUL_S(a, x);
+
+	return XT_TRUNC_S(y, 0);
+}
+
+/* Fractional multiplication with shift and round
+ * Note that the parameters px and py must be cast to (int64_t) if other type.
+ */
+static inline int exp_hifi_q_multsr_32x32(int a, int b, int c, int d, int e)
```
Comment (singalsu) [inline src/math/exp_fcn_hifi.c:303]: This is also suspicious, there's no need for generic C speed optimization with generic fractions. You should do this with inline instrinsics with minimal operations if need fast intrinsic multiply. It's would be usually just an AE multiply instruction with a literal constant shift to adjust the fractional format. Please leave optimizations like this to other PR where we can review them more carefully.

File: src/math/exp_fcn_hifi.c [hunk 21]
```diff
@@ -277,10 +280,72 @@ int32_t sofm_exp_int32(int32_t x)
...
+	res = AE_SRAA64(res, shift);
+	res = AE_ADD64(res, 1);
+	res = AE_SRAI64(res, 1);
+	xt_o = AE_MOVINT32_FROMINT64(res);
+
+	return xt_o;
+}
+
+/* A macro for Q-shifts */
+static inline int exp_hifi_q_shift_rnd(int a, int b, int c)
```
Comment (singalsu) [inline src/math/exp_fcn_hifi.c:320]: Same for all other fractional C macros replacements.

**Comment by singalsu** [comment]
Seems it's not ensured by C standards that literals floating point macros are calculated in the C pre-processor. Maybe things have changed with xt-clang.

https://stackoverflow.com/questions/21241031/does-the-c-preprocessor-handle-floating-point-math-constants

It would be safest to change macros like `#define SOME_COEF Q_CONVERT_FLOAT(-11.5, 27)` into `#define SOME_COEF -1543503872 /* -11.5 Q5.27 */`.

**Comment by singalsu** [comment]
> Seems it's not ensured by C standards that literals floating point macros are calculated in the C pre-processor. Maybe things have changed with xt-clang.
> 
> https://stackoverflow.com/questions/21241031/does-the-c-preprocessor-handle-floating-point-math-constants
> 
> It would be safest to change macros like `#define SOME_COEF Q_CONVERT_FLOAT(-11.5, 27)` into `#define SOME_COEF -1543503872 /* -11.5 Q5.27 */`.

@ShriramShastry Please check disassembly of your Q_CONVERT_FLOAT() usage. We looked into TGL and MTL builds with @btian1 and we could not find floating point code generation from current usages of the macro. No, float overhead found. So, let's not yet start to replace these before understanding better as I worried in my previous comment.

File: src/math/exp_fcn_hifi.c (hunk 12 above)
Comment (btian1) [reply src/math/exp_fcn_hifi.c:367]: I mean why not directly use:Q_CONVERT_FLOAT(-11.5, 27)?

File: test/cmocka/src/math/window/CMakeLists.txt [hunk 22]
```diff
@@ -8,4 +8,5 @@ cmocka_test(window
 	${PROJECT_SOURCE_DIR}/src/math/base2log.c
 	${PROJECT_SOURCE_DIR}/src/math/decibels.c
 	${PROJECT_SOURCE_DIR}/src/math/exp_fcn.c
+	${PROJECT_SOURCE_DIR}/src/math/exp_fcn_hifi.c
```
Comment (btian1) [inline test/cmocka/src/math/window/CMakeLists.txt:11]: please squash this patch with previous one, as Marc said, you need get each patch pass all CI separately, here, there is obvious CI error.

**Review by singalsu — APPROVED** [review]

File: src/include/sof/math/exp_fcn.h [hunk 23]
```diff
@@ -26,6 +26,38 @@
 
 #endif
 
-int32_t sofm_exp_int32(int32_t x);
+/* TODO: Is there a MCPS difference */
+#define USING_QCONVERT	1
```
Comment (singalsu) [inline src/include/sof/math/exp_fcn.h:30]: I already approved, but let's check in real devices if setting this to zero speeds up the code. In theory it shouldn't impact since Q_CONVERT_FLOAT macro should be evaluated in C pre-processor. Once confirmed we can remove the direct integers. Or if difference seen, remove the Q_CONVERT_FLOAT part.

Comment (lgirdwood) [reply src/include/sof/math/exp_fcn.h:30]: ok, can we do this as next steps in a follow up PR and test drive this now.

**Review by lgirdwood — APPROVED** [review]
